chrome/content/zotero/xpcom/connector/server_connector.js

```diff
@@ -1843,3 +1843,93 @@ Zotero.Server.Connector.IEHack.prototype = {
 			'</head><body></body></html>');
 	}
 }
+
+/**
+ * Make an HTTP request from the client. Accepts {@link Zotero.HTTP.request} options and returns a minimal response
+ * object with the same form as the one returned from {@link Zotero.Utilities.Translate#request}.
+ *
+ * Accepts:
+ *		method - The request method ('GET', 'POST', etc.)
+ *		url - The URL to make the request to. Must be an absolute HTTP(S) URL.
+ *		options - See Zotero.HTTP.request() documentation. Differences:
+ *			- responseType is always set to 'text'
+ *			- successCodes is always set to false (non-2xx status codes will not trigger an error)
+ * Returns:
+ *		Response code is always 200. Body contains:
+ *			status - The response status code, as a number
+ *			headers - An object mapping header names to values
+ *			body - The response body, as a string
+ */
+Zotero.Server.Connector.Request = function () {};
+Zotero.Server.Endpoints["/connector/request"] = Zotero.Server.Connector.Request;
+Zotero.Server.Connector.Request.prototype = {
+	supportedMethods: ["POST"],
+	supportedDataTypes: ["application/json"],
+
+	/**
+	 * For testing: allow disabling URL validation so we can make requests to the server.
+	 */
+	validateURLs: true,
+
+	init: async function (req) {
+		let { method, url, options } = req.data;
+		
+		if (typeof method !== 'string' || typeof url !== 'string') {
+			return [400, 'text/plain', 'method and url are required and must be strings'];
+		}
+		
+		let uri;
+		try {
+			uri = Services.io.newURI(url);
+		}
+		catch (e) {
+			return [400, 'text/plain', 'Invalid URL'];
+		}
+		
+		if (uri.scheme != 'http' && uri.scheme != 'https') {
+			return [400, 'text/plain', 'Unsupported scheme'];
+		}
+
+		// Through heuristics, attempt to prevent:
+		//   - recursive requests to this endpoint, which will deadlock the server until they time out
+		//   - requests to the local network, which shouldn't be necessary from the connector and pose a security risk
+		if (this.validateURLs
+			&& (uri.port == Zotero.Prefs.get('httpServer.port')
+				|| uri.filePath == '/connector/request'
+				|| Services.io.hostnameIsLocalIPAddress(uri))) {
```

Comment:
Seems like all of this could just be replaced with blocking `localhost` or any IP address (which would include local addresses and loopback), but I would even say we should limit this to just the domain(s) we need for now. We've never needed this before, and hopefully won't need it for anything else, so I'd rather not unnecessarily expose an open HTTP relay. We should also check explicitly that User-Agent starts with `Mozilla/`, which will guarantee that it's a request with a connector header (since that's checked in server.js).

Basically, if Zotero is specifically given network access on an otherwise locked-down system, we don't want it to be usable for a privilege escalation attack.

Reply:
(I guess checking for the path would prevent a recursive request if the computer somehow had a non-local address, though that would be pretty rare. In any case, I think better just to allow specific domains.)